server/internal/router/router.go

```diff
@@ -0,0 +1,184 @@
+	if err != nil {
+		return err
+	}
+
+	if err := is.SetPodEventHandlers(ctx, []k8s.EventHandlerWithContext{
+		NewPodEventHandler(r.m, errCh),
+	}); err != nil {
+		return err
+	}
+
+	is.Start()
+	log.Printf("Pod informer started.\n")
+
+	// refresh routes periodically, since a engine may evict models.
+	ticker := time.NewTicker(tickPeriod)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ticker.C:
+			if err := r.refreshRoutes(ctx, r.imeConfig.Namespace, engineLabels); err != nil {
+				return err
+			}
+			r.m.printRoute()
+		case <-stopper:
+			log.Printf("run done.\n")
+			return nil
+		case err := <-errCh:
+			log.Printf("run error: %s", err)
+			return err
+		}
+	}
+}
+
+// GetEngineForModel returns the engine address (ip and port) for the given model.
+func (r *Router) GetEngineForModel(ctx context.Context, modelID string) (string, error) {
+	routes := r.m.getRoute(model{id: modelID})
+	if len(routes) != 0 {
+		// TODO(guangrui): handle load balance.
+		return routes[0], nil
+	}
+	ip, err := r.pullModel(ctx, modelID)
+	if err != nil {
+		return "", err
+	}
+	return fmt.Sprintf("%s:%d", ip, r.imeConfig.OllamaPort), nil
+}
+
+func (r *Router) refreshRoutes(ctx context.Context, namespace string, labels map[string]string) error {
+	k8sClient, err := k8s.NewClient()
+	if err != nil {
+		return fmt.Errorf("create K8sClient: %v", err)
+	}
+
+	pods, err := k8sClient.ListPods(ctx, namespace, labels)
+	if err != nil {
+		return err
+	}
+	log.Printf("Refreshing routes, found %d pods\n", len(pods))
+
+	newRoute := newRoute()
+	for _, pod := range pods {
+		ip := pod.Status.PodIP
+		if ip == "" {
+			continue
+		}
+		models, err := r.listModels(ctx, ip)
+		if err != nil {
+			return err
+		}
+		for _, m := range models {
+			newRoute.addRoute(model{id: m}, ip)
+		}
+		newRoute.addServer(ip)
+	}
+	newRoute.printRoute()
+
+	r.m.replace(newRoute)
+	return nil
+}
+
+func (r *Router) listModels(ctx context.Context, ip string) ([]string, error) {
+	dest := fmt.Sprintf("%s:%d", ip, r.imeConfig.InternalGRPCPort)
+	log.Printf("Listing loaded models on %s\n", dest)
+	conn, err := grpc.Dial(dest, grpc.WithTransportCredentials(insecure.NewCredentials()))
+	if err != nil {
+		return nil, err
+	}
+	ec := v1.NewInferenceEngineInternalServiceClient(conn)
+	resp, err := ec.ListModels(ctx, &v1.ListModelsRequest{})
+	if err != nil {
+		return nil, err
+	}
+
+	var modelIDs []string
+	for _, model := range resp.Models {
+		modelIDs = append(modelIDs, model.Id)
+	}
+	return modelIDs, nil
+}
+
+// pullModel pulls the model to the least loaded engine and returns the IP address of the engine.
+func (r *Router) pullModel(ctx context.Context, modelID string) (string, error) {
+	ip, err := r.m.findLeastLoadedEngine()
+	if err != nil {
+		return "", err
+	}
+
+	dest := fmt.Sprintf("%s:%d", ip, r.imeConfig.InternalGRPCPort)
+	log.Printf("Pulling model %s to %s\n", modelID, dest)
+	conn, err := grpc.Dial(dest, grpc.WithTransportCredentials(insecure.NewCredentials()))
```

Comment:
We probably want to define a factory for this and inject to this struct so that we can fake/mock in the unit test.